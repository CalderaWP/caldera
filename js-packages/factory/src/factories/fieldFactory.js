import {
	InputField,
	SelectField,
	isValidHtml5type,
	FieldSet,
	Message
} from '@caldera-labs/components';
import React, {Fragment} from 'react';


const TheMessage = (messages, fieldId) => {
	return undefined !== typeof messages && messages.hasOwnProperty('message') ? <Message
		message={{
			message: 'Hi Roy',
			error: false,
		}}
	/> : <Fragment/>;
}
export const fieldFactory = (field, onChange, onBlur) => {
	const {fieldType, label, attributes, options, fieldId, messages} = field;

	const Message = TheMessage(messages, fieldId);
	switch (fieldType) {
		case 'checkbox':
		case 'checkboxes' :
			if ('checkboxes' === fieldType || fieldType.hasOwnProperty('options')) {
				return (
					<FieldSet
						fieldType={fieldType}
						legend={label}
						attributes={attributes}
					>
						{options.map(option => {
							const {
								value,
								label,
								description,
								attributes
							} = option;
							const fieldId = option.hasOwnProperty('id') ? option.id : `opt-${fieldId}-${option.value}`;
							return <InputField
								key={fieldId}
								fieldId={fieldId}
								value={value}
								label={label}
								description={description}
								html5type={'checkbox'}
								attributes={attributes}
								onChange={onChange}
							/>
						})}
						<Message/>
					</FieldSet>
				);
			} else {
				return <InputField
					{...field} onChange={onChange}
				><TheMessage/></InputField>
			}
			break;
		case 'select':
		case 'dropdown':
			return <SelectField
				{...field}
				onChange={onChange}

			><Message/></SelectField>;
			break;
		case 'text':
		case 'email':
		case 'number':
		case 'input':
		default:
			if (isValidHtml5type(fieldType)) {
				field.html5type = fieldType;
			} else {
				field.html5type = 'text';

			}
			return <InputField
				{...field}
				onChange={onChange}
			><TheMessage/></InputField>
			break;
	}
}
