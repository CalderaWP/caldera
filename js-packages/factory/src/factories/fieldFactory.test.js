import {fieldFactory}  from './fieldFactory';
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from '../fields.fixtures';

describe( 'fieldFactory', () => {
	let onChange;
	let onBlur;
	beforeEach( () => {
		onChange= jest.fn();
		onBlur= jest.fn()
	});
	it( 'Creates a text field', () => {
		const component = renderer.create(fieldFactory(textField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates a text field when field type is bogus', () => {
		const component = mount(fieldFactory({...textField,fieldType:'travels-in-space'},onChange,onBlur));
		expect(component.find( 'input').prop('type')).toBe('text');
	});
	it( 'Creates a text field when field type is input', () => {
		const component = mount(fieldFactory({...textField,fieldType:'input'},onChange,onBlur));
		expect(component.find( 'input').prop('type')).toBe('text');
	});

	it( 'Creates a number field', () => {
		const component = renderer.create(fieldFactory(numberField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates an email field', () => {
		const component = renderer.create(fieldFactory(emailField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates an checkbox field field', () => {
		const component = renderer.create(fieldFactory(checkboxField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates an checkbox field set', () => {
		const component = renderer.create(fieldFactory(checkboxFieldset,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates an select field', () => {
		const component = renderer.create(fieldFactory(selectField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates a radio field', () => {
		const component = renderer.create(fieldFactory(radioField,onChange,onBlur));
		expect(component.toJSON()).toMatchSnapshot();
	});

	it( 'Creates an select field identified as dropdown', () => {

		const component = renderer.create(fieldFactory({
			...selectField,
			fieldType:'dropdown'
		}),onChange,onBlur);
		expect(component.toJSON()).toMatchSnapshot();
	});
});
