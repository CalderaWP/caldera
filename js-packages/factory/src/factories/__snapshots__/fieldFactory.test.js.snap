// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldFactory Creates a checkbox fieldset 1`] = `
<fieldset
  className="caldera-fieldset caldera-fieldset-checkboxes"
>
  <legend>
    Checkbox Field Set Label
  </legend>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-1"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-4"
      >
        One
      </label>
    </div>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby="inspector-checkbox-control-5__help"
        checked={true}
        className="components-checkbox-control__input"
        id="opt-2"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-5"
      >
        Two
      </label>
    </div>
    <p
      className="components-base-control__help"
      id="inspector-checkbox-control-5__help"
    >
      The Second Option
    </p>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-3"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-6"
      >
        Three
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`fieldFactory Creates a number field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-number"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-3"
    >
      Number Field Label
    </label>
    <input
      aria-describedby="inspector-text-control-3__help"
      className="components-text-control__input"
      id="numberFieldId"
      max={10}
      min={1}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="number"
      value={3}
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-3__help"
  >
    Number field description
  </p>
</div>
`;

exports[`fieldFactory Creates a radio field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-radio components-radio-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-radio-control-0"
    >
      Radio Field Label
    </label>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={false}
        className="components-radio-control__input"
        id="inspector-radio-control-0-0"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={0}
      />
      <label
        htmlFor="inspector-radio-control-0-0"
      >
        Zero
      </label>
    </div>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={true}
        className="components-radio-control__input"
        id="inspector-radio-control-0-1"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        htmlFor="inspector-radio-control-0-1"
      >
        One
      </label>
    </div>
    <div
      className="components-radio-control__option"
    >
      <input
        aria-describedby="inspector-radio-control-0__help"
        checked={false}
        className="components-radio-control__input"
        id="inspector-radio-control-0-2"
        name="inspector-radio-control-0"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        htmlFor="inspector-radio-control-0-2"
      >
        Two
      </label>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-radio-control-0__help"
  >
    Radio field description
  </p>
</div>
`;

exports[`fieldFactory Creates a text field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-text"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-0"
    >
      First Name
    </label>
    <input
      aria-describedby="inspector-text-control-0__help"
      className="components-text-control__input"
      id="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="text"
      value="roy"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-0__help"
  >
    your first name
  </p>
</div>
`;

exports[`fieldFactory Creates an checkbox field field 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <input
      aria-describedby="inspector-checkbox-control-0__help"
      checked={true}
      className="components-checkbox-control__input"
      id="checkboxFieldLabel"
      onBlur={[Function]}
      onChange={[Function]}
      type="checkbox"
      value="1"
    />
    <label
      className="components-checkbox-control__label"
      htmlFor="inspector-checkbox-control-0"
    >
      Checkbox Labe
    </label>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-checkbox-control-0__help"
  >
    Checkbox field description
  </p>
</div>
`;

exports[`fieldFactory Creates an checkbox field set 1`] = `
<fieldset
  className="caldera-fieldset caldera-fieldset-checkboxes"
>
  <legend>
    Checkbox Field Set Label
  </legend>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-1"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-1"
      >
        One
      </label>
    </div>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby="inspector-checkbox-control-2__help"
        checked={true}
        className="components-checkbox-control__input"
        id="opt-2"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-2"
      >
        Two
      </label>
    </div>
    <p
      className="components-base-control__help"
      id="inspector-checkbox-control-2__help"
    >
      The Second Option
    </p>
  </div>
  <div
    className="components-base-control"
  >
    <div
      className="components-base-control__field"
    >
      <input
        aria-describedby={undefined}
        checked={true}
        className="components-checkbox-control__input"
        id="opt-3"
        onBlur={[Function]}
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <label
        className="components-checkbox-control__label"
        htmlFor="inspector-checkbox-control-3"
      >
        Three
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`fieldFactory Creates an email field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-email"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-text-control-4"
    >
      Email Field Label
    </label>
    <input
      aria-describedby="inspector-text-control-4__help"
      className="components-text-control__input"
      id="emailFieldId"
      multiple={true}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder={undefined}
      type="email"
      value="emai@email.com"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-text-control-4__help"
  >
    Email field description
  </p>
</div>
`;

exports[`fieldFactory Creates an select field 1`] = `
<div
  className="components-base-control caldera-field caldera-field-select"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-select-control-0"
    >
      Select Field Label
    </label>
    <select
      aria-describedby="inspector-select-control-0__help"
      className="components-select-control__input"
      id="selectFieldId"
      multiple={false}
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        Zero
      </option>
      <option
        value={1}
      >
        One
      </option>
      <option
        value={2}
      >
        Two
      </option>
    </select>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-select-control-0__help"
  >
    Select field description
  </p>
</div>
`;

exports[`fieldFactory Creates an select field identified as dropdown 1`] = `
<div
  className="components-base-control caldera-field caldera-field-select"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-select-control-1"
    >
      Select Field Label
    </label>
    <select
      aria-describedby="inspector-select-control-1__help"
      className="components-select-control__input"
      id="selectFieldId"
      multiple={false}
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        Zero
      </option>
      <option
        value={1}
      >
        One
      </option>
      <option
        value={2}
      >
        Two
      </option>
    </select>
  </div>
  <p
    className="components-base-control__help"
    id="inspector-select-control-1__help"
  >
    Select field description
  </p>
</div>
`;

exports[`fieldFactory should create textarea 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="inspector-textarea-control-0"
    >
      Your message
    </label>
    <textarea
      aria-describedby="inspector-textarea-control-0__help"
      className="components-textarea-control__input"
      cols={8}
      id="textArea1"
      maxLength={500}
      minlength={10}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      rows={5}
      value="The content of text area"
    />
  </div>
  <p
    className="components-base-control__help"
    id="inspector-textarea-control-0__help"
  >
    your first name
  </p>
</div>
`;

exports[`fieldFactory should create toggle 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="toggleFieldLabel"
    >
      Toggle Label
    </label>
    <div
      className="components-base-control components-toggle-control"
    >
      <div
        className="components-base-control__field"
      >
        <span
          className="components-form-toggle is-checked"
        >
          <input
            aria-describedby={undefined}
            checked={true}
            className="components-form-toggle__input"
            id="inspector-toggle-control-0"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="components-form-toggle__track"
          />
          <span
            className="components-form-toggle__thumb"
          />
          <svg
            aria-hidden="true"
            className="components-form-toggle__on"
            focusable="false"
            height="6"
            role="img"
            viewBox="0 0 2 6"
            width="2"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h2v6H0z"
            />
          </svg>
        </span>
        <label
          className="components-toggle-control__label"
          htmlFor="inspector-toggle-control-0"
        />
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="toggleFieldLabel__help"
  >
    Toggle field description
  </p>
</div>
`;
